File: python/mr_x_and_his_shots.py
def solve(shots, players):
    n = len(shots)
    m = len(players)
    
    # Extract shot start and end points
    starts = [shot[0] for shot in shots]
    ends = [shot[1] for shot in shots]
    
    # Sort for binary search
    starts.sort()
    ends.sort()
    
    total_strength = 0
    
    for player in players:
        C = player[0]
        D = player[1]
        
        # Count shots with start <= D
        count_start_before_end = bisect.bisect_right(starts, D)
        
        # Count shots with end < C
        count_end_before_start = bisect.bisect_left(ends, C)
        
        # Overlapping shots
        overlap = count_start_before_end - count_end_before_start
        total_strength += overlap
    
    return total_strength

File: javaScript/mr_x_and_his_shots.js
function binarySearchRight(arr, target) {
    let left = 0;
    let right = arr.length;
    
    while (left < right) {
        const mid = Math.floor((left + right) / 2);
        if (arr[mid] <= target) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    return left;
}

function binarySearchLeft(arr, target) {
    let left = 0;
    let right = arr.length;
    
    while (left < right) {
        const mid = Math.floor((left + right) / 2);
        if (arr[mid] < target) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    return left;
}

function solve(shots, players) {
    const n = shots.length;
    const m = players.length;
    
    // Extract shot start and end points
    const starts = new Array(n);
    const ends = new Array(n);
    
    for (let i = 0; i < n; i++) {
        starts[i] = shots[i][0];
        ends[i] = shots[i][1];
    }
    
    // Sort for binary search
    starts.sort((a, b) => a - b);
    ends.sort((a, b) => a - b);
    
    let totalStrength = 0;
    
    for (let i = 0; i < m; i++) {
        const C = players[i][0];
        const D = players[i][1];
        
        // Count shots with start <= D
        const countStartBeforeEnd = binarySearchRight(starts, D);
        
        // Count shots with end < C
        const countEndBeforeStart = binarySearchLeft(ends, C);
        
        // Overlapping shots
        const overlap = countStartBeforeEnd - countEndBeforeStart;
        totalStrength += overlap;
    }
    
    return totalStrength;
}